gillespy2/solvers/cpp/c_base/tau_leaping_cpp_solver/tau_leaper.cpp

```diff
@@ -0,0 +1,330 @@
+    double tau; //tau time to step;
+    std::map<std::string,double> critical_taus;    //Mapping of possible critical_taus, to be evaluated
+    std::map<std::string,double> mu_i;
+    std::map<std::string,double> sigma_i;
+    bool critical = false;  // system-wide flag, true when any reaction is critical
+    double non_critical_tau = 0;  // holds the smallest tau time for non-critical reactions
+    double critical_tau = 0;  // holds the smallest tau time for critical reactions
+
+    int v;//used for number of population reactant consumes
+
+    // initialize mu_i and sigma_i to 0
+    for(int spec = 0; spec<model.number_species; spec++){
+            mu_i[model.species[spec].name]= 0;
+            sigma_i[model.species[spec].name] = 0;
+
+        }
+    // Determine if there are any critical reactions, update mu_i and sigma_i
+    for (int reaction = 0; reaction < model.number_reactions; reaction++){
+        for (auto const &reactant: tau_args.reactions_reactants[reaction]){
+            if (model.reactions[reaction].species_change[reactant]<0){
+                v = abs(model.reactions[reaction].species_change[reactant]);
+
+                if ((double)current_state[reactant] / v < tau_args.critical_threshold && propensity_values[reaction] > 0){
+                    critical = true; // Critical reaction present in simulation
+                }
+                int consumed = abs(model.reactions[reaction].species_change[reactant]);
+                mu_i[model.species[reactant].name] += consumed*propensity_values[reaction];//Cao, Gillespie, Petzold 32a
+                sigma_i[model.species[reactant].name] += std::pow(consumed,2) * propensity_values[reaction];//Cao, Gillespie, Petzold 32a
+            }
+        }
+    }
+
+    // If a critical reaction is present, estimate tau for a single firing of each
+    // critical reaction with propensity > 0, and take the smallest tau
+    if (critical == true){
+        for (int reaction = 0; reaction < model.number_reactions; reaction++){
+            if (propensity_values[reaction]>0)
+                critical_taus[model.reactions[reaction].name] = 1/propensity_values[reaction];
+        }
+        std::pair<std::string, double> min;
+        //find min of critical_taus
+        min = *min_element(critical_taus.begin(), critical_taus.end(),[](const auto& lhs, const auto& rhs){ return lhs.second < rhs.second;});
+        critical_tau = min.second;
+    }
+
+    if (tau_args.g_i_lambdas.size()>0){
+        for (auto const& x : tau_args.g_i_lambdas)
+           {
+               tau_args.g_i[x.first] = tau_args.g_i_lambdas[x.first](tau_args.g_i[x.first]);
+               tau_args.epsilon_i[x.first] = tau_tol / tau_args.g_i[x.first];
+               tau_args.g_i_lambdas.erase(x.first);//MAYBE SHOULDN'T ERASE, may break loop/may be understanding this part wrong
```

Comment:
Regarding code comment - Is this safe / has this been tested?

Reply:
Yes, this had to do with something else I was testing, I seem to have left the comment in -- I'll delete it